compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -922,6 +922,47 @@ bool OMR::ValuePropagation::transformUnsafeCopyMemoryCall(TR::Node *arrayCopyNod
    return false;
    }
 
+static void setNodeOnList(TR::Node *currNode, TR_BitVector *nodeList)
+   {
+   if (nodeList->isSet(currNode->getGlobalIndex()))
+      return;
+
+   nodeList->set(currNode->getGlobalIndex());
+
+   for (int32_t i = 0; i < currNode->getNumChildren(); i++)
+      {
+      setNodeOnList(currNode->getChild(i), nodeList);
+      }
+   }
+
+static void findAndReplaceCommonedNodeWithLoadFromTemp(TR::Node *currNode, TR::Node *matchingNode, TR::SymbolReference *symRef, TR_BitVector *nodeSkipList, bool trace, TR::Compilation *comp)
+   {
+   for (int32_t i = 0; i < currNode->getNumChildren(); i++)
+      {
+      TR::Node *childNode = currNode->getChild(i);
+
+      if ((childNode == matchingNode)
+          || ((childNode->getOpCodeValue() == TR::aload) &&
+              (matchingNode->getOpCodeValue() == TR::aload) &&
+              childNode->getOpCode().hasSymbolReference() &&
+              matchingNode->getOpCode().hasSymbolReference() &&
+              (childNode->getSymbolReference()->getReferenceNumber() == matchingNode->getSymbolReference()->getReferenceNumber())))
+         {
+         TR::Node *loadNode = TR::Node::createLoad(symRef);
+
+         childNode->recursivelyDecReferenceCount();
+         currNode->setAndIncChild(i, loadNode);
+
+         if (trace)
+            traceMsg(comp, "%s: currNode n%dn replace childNode n%dn with loadNode n%dn\n", __FUNCTION__, currNode->getGlobalIndex(), childNode->getGlobalIndex(), loadNode->getGlobalIndex());
+         }
+      else if (!nodeSkipList->isSet(childNode->getGlobalIndex()))
```

Comment:
I find it a little bit confusing that a node that is in `nodeSkipList` might get processed by the `if` that appears just before this `else if` if its parent was not in `nodeSkipList`.  I wonder if it should be called `skipDescendantsList` or something like that.  Just a thought.